Std/Tactic/Simps/Basic.lean

```diff
@@ -0,0 +1,1157 @@
+
+/-- Function elaborating `initialize_simps_projections`. -/
+@[command_elab «initialize_simps_projections»] def elabInitializeSimpsProjections : CommandElab
+  | stx@`(initialize_simps_projections $[?%$trc]? $id $[($stxs,*)]?) => do
+    let stxs := stxs.getD <| .mk #[]
+    let rules ← stxs.getElems.raw.mapM elabSimpsRule
+    let nm ← resolveGlobalConstNoOverload id
+    _ ← liftTermElabM <| addTermInfo id.raw <| ← mkConstWithLevelParams nm
+    _ ← liftCoreM <| getRawProjections stx nm true rules trc.isSome
+  | _ => throwUnsupportedSyntax
+
+/-- Configuration options for `@[simps]` -/
+structure Config where
+  /-- Make generated lemmas simp lemmas -/
+  isSimp := true
+  /-- Other simp-attributes to apply to generated lemmas.
+  Attributes that are currently not simp-attributes are not supported. -/
+  attrs : List Name := []
+  /-- simplify the right-hand side of generated simp-lemmas using `dsimp, simp`. -/
+  simpRhs := false
+  /-- TransparencyMode used to reduce the type in order to detect whether it is a structure. -/
+  typeMd := TransparencyMode.instances
+  /-- TransparencyMode used to reduce the right-hand side in order to detect whether it is a
+  constructor. Note: was `none` in Lean 3 -/
+  rhsMd := TransparencyMode.reducible
+  /-- Generated lemmas that are fully applied, i.e. generates equalities between applied functions.
+  Set this to `false` to generate equalities between functions. -/
+  fullyApplied := true
+  /-- List of types in which we are not recursing to generate simplification lemmas.
+  E.g. if we write `@[simps] def e : α × β ≃ β × α := ...` we will generate `e_apply` and not
+  `e_apply_fst`. -/
+  notRecursive := [`Prod, `PProd, `Opposite, `PreOpposite]
+  /-- Output debug messages. Not used much, use `set_option simps.debug true` instead. -/
+  debug := false
+  deriving Inhabited
+
+/-- Function elaborating `Config` -/
+declare_config_elab elabSimpsConfig Config
+
+/-- A common configuration for `@[simps]`: generate equalities between functions instead equalities
+  between fully applied Expressions. Use this using `@[simps (config := .asFn)]`. -/
+def Config.asFn : Simps.Config where
+  fullyApplied := false
+
+/-- A common configuration for `@[simps]`: don't tag the generated lemmas with `@[simp]`.
+  Use this using `@[simps (config := .lemmasOnly)]`. -/
+def Config.lemmasOnly : Config where
+  isSimp := false
+
+/-- `instantiateLambdasOrApps es e` instantiates lambdas in `e` by expressions from `es`.
+If the length of `es` is larger than the number of lambdas in `e`,
+then the term is applied to the remaining terms.
+Also reduces head let-expressions in `e`, including those after instantiating all lambdas.
+
+This is very similar to `expr.substs`, but this also reduces head let-expressions. -/
+partial def _root_.Lean.Expr.instantiateLambdasOrApps (es : Array Expr) (e : Expr) : Expr :=
+  e.betaRev es.reverse true -- check if this is what I want
```

Comment:
Comment seems fishy and this should either be moved to `Std/Lean/Expr.lean` or renamed.

Reply:
I never tested whether this function satisfies the docstring exactly, and it doesn't do the second half of ``Also reduces head let-expressions in `e`, including those after instantiating all lambdas.``

In https://github.com/leanprover-community/mathlib4/compare/master...simps_upstream_tests I tried to see if doing that breaks anything in Mathlib. I expect not, given the low number of `let` expressions in definitions. 

We can also keep the current functionality. It would be nice if `Lean.Expr.beta` supports the optional flags of `Lean.Expr.betaRev`, then this function doesn't need to exist.